module/errorhandler/erroraware.go

```diff
@@ -0,0 +1,165 @@
+// including in interfaces that compose it.
+type IrrecoverableSignalerContext interface {
+	context.Context
+	ThrowIrrecoverable(err error) // delegates to the signaler
+	sealed()                      // private, to constrain builder to using WithIrrecoverableSignal
+}
+
+// private, to force context derivation / WithIrrecoverableSignal
+type irrecoverableSignalerCtxt struct {
+	context.Context
+	signaler *IrrecoverableSignaler
+}
+
+func (irs irrecoverableSignalerCtxt) sealed() {}
+
+// Drop-in replacement for panic, log.Fatal, log.Panic, etc
+// to use when we are able to get an IrrecoverableSignalerContext and thread it down in the component
+func (irs irrecoverableSignalerCtxt) ThrowIrrecoverable(err error) {
+	irs.signaler.ThrowIrrecoverable(err)
+}
+
+// the One True Way of getting an IrrecoverableSignalerContext
+func WithIrrecoverableSignal(ctx context.Context, sig *IrrecoverableSignaler) IrrecoverableSignalerContext {
+	return irrecoverableSignalerCtxt{ctx, sig}
+}
+
+// If we have an IrrecoverableSignalerContext, we can directly ctx.ThrowIrrecoverable.
+//
+// But a lot of library methods expect context.Context, & we want to pass the same w/o boilerplate
+// Moreover, we could have built with: context.WithCancel(context.WithIrrecoverableSignal(ctx, sig), ...)
+// "downcasting" to context.Context. Yet, we can still type-assert and recover.
+//
+// ThrowIrrecoverable can be a drop-in replacement anywhere we have a context.Context likely
+// to support Irrecoverables. Note: this is not a method
+func ThrowIrrecoverable(ctx context.Context, err error) {
+	signalerAbleContext, ok := ctx.(irrecoverableSignalerCtxt)
+	if ok {
+		signalerAbleContext.signaler.ThrowIrrecoverable(err)
+	}
+	// Be spectacular on how this does not -but should- handle irrecoverables:
+	log.Fatalf("Irrecoverable error signaler not found for context, please implement! Unhandled irrecoverable error %v", err)
+}
+
+////////////////////////////////////////////////
+// Integrating it w/ ReadyDoneAware & friends //
+////////////////////////////////////////////////
+
+// If we want to do it using interface composition (see module.Component), we will need to build on the existing intrefaces
+// note the irrecoverable management needs to be:
+// - set up before the call to start (if the start itself meets an irrecoverable condition)
+// - not throw / return error itself, except to an enclosing context
+
+type Startable interface {
+	Start(runCtx IrrecoverableSignalerContext) error
+}
+
+// later, in RunComponent, the plumbing happens ...
+type Component interface {
+	Startable
+	module.ReadyDoneAware
+}
+type ComponentFactory func() (Component, error)
+
+func RunComponent(parentCtx IrrecoverableSignalerContext, componentFactory ComponentFactory, handler IrrecoverableHandler) {
+
+	component, err := componentFactory()
+	if err != nil {
+		parentCtx.ThrowIrrecoverable(err) // failure to generate the component
+	}
```

Comment:
Based on my understanding of this code, I don't think `parentCtx.ThrowIrrecoverable` is a good way to handle this error.

Because there is no way to directly handle errors on `IrrecoverableSignaler`, the only way this error eventually gets handled is *by the `IrrecoverableHandler` for the parent component*, which is logic that is owned by the grandparent component. I don't think it makes sense that a failure to start a child component should be something that is handled by the grandparent component. The grandparent component does not even have any way to know which child component failed.

We should either implement a way to directly handle errors on `IrrecoverableSignaler` without using `RunComponent`, or we should directly return the error here.

Reply:
Fair enough: I just want to exemplify that this is a failure to start, and so it should go outside of the normal band that deals with irrecoverables through restarts: restarting here would only create the same error.

Hence, I think the function we use as a continuation (`restart`) which has the signature `func ()` should have `func () error.